### PR TITLE
Network Explorer

### PR DESCRIPTION
**Pull request by mjohnson541**
This PR adds functionality for loading the database within Cantherm, using RMG to expand Cantherm networks and reduction of the expanded networks.

### CODE REVIEW

**Comment by mjohnson541** [comment]
Tests are failing because the RMG server is down

File: rmgpy/rmg/pdep.py [hunk 1]
```diff
@@ -318,6 +321,194 @@ def addPathReaction(self, newReaction):
...
+        reactionModel.updateUnimolecularReactionNetworks()
+        
+        if reactionModel.pressureDependence.outputFile:
+            path = os.path.join(reactionModel.pressureDependence.outputFile,'pdep')
+            
+            for name in os.listdir(path): #remove the old reduced file
+                if name.endswith('reduced.py'):
+                    os.remove(os.path.join(path,name))
+                    
+            for name in os.listdir(path): #find the file that has
```
Comment (alongd) [inline rmgpy/rmg/pdep.py]: incomplete comment

File: rmgpy/cantherm/explorer.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        self.pdepjob.execute(outputFile, plot, format='pdf', print_summary=True)
+        
+        if warns != []:
+            logging.info('\nOUTPUT WARNINGS:\n')
+            for w in warns:
+                logging.warning(w)
+        
+        
+        
+
```
Comment (alongd) [inline rmgpy/cantherm/explorer.py]: empty line at end of file?
(and remove extra tabs?)

File: rmgpy/cantherm/explorer.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            if not os.path.exists(os.path.join(reactionModel.pressureDependence.outputFile,'pdep')):
+                os.mkdir(os.path.join(reactionModel.pressureDependence.outputFile,'pdep'))
+            else:
+                shutil.rmtree(os.path.join(reactionModel.pressureDependence.outputFile,'pdep'))
+                os.mkdir(os.path.join(reactionModel.pressureDependence.outputFile,'pdep'))
+            
+        #get the molecular formula for the network
+        
+        mmol = None
+        for spc in self.source:
```
Comment (alongd) [inline rmgpy/cantherm/explorer.py]: Consider using a shorter syntax instead:
```python
if len(self.source) == 1:
    form = self.source[0].molecule[0].getFormula()
else:
    form = self.source[0].molecule[0].merge(self.source[1].molecule[0]).getFormula()
```
Also, the formula can be generated from any isomer instead

File: rmgpy/cantherm/explorer.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                      bimolecularReact=np.zeros((len(reactionModel.core.species),len(reactionModel.core.species))))
+        
+        #find the network we're interested in
+        for nwk in reactionModel.networkList:
+            if set(nwk.source) == set(self.source):
+                self.source = nwk.source
+                network = nwk
+                break
+        else:
+            raise ValueError('did not generate a network with the requested source')
```
Comment (alongd) [inline rmgpy/cantherm/explorer.py]: Use `NetworkError` instead of `ValueError`

File: rmgpy/cantherm/explorer.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        
+        explore_tol = self.explore_tol
+        
+        incomplete = True
+        
+        while incomplete:
+            incomplete = False
+            for T in Tlist:
+                for P in Plist:
+                    if network.getLeakCoefficient(T=T,P=P) > explore_tol:
```
Comment (alongd) [inline rmgpy/cantherm/explorer.py]: `explore_tol` is only used here. Perhaps use `self.explore_tol` and remove the above ` explore_tol = self.explore_tol`?

File: rmgpy/cantherm/explorer.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        for spc in reactionModel.core.species:
+            for i,item in enumerate(self.source):
+                if spc.isIsomorphic(item):
+                    self.source[i] = spc
+        
+        
+        #react initial species
+        flags = np.array([s.molecule[0].getFormula()==form for s in reactionModel.core.species])
+        logging.info('flags: {0}'.format(flags))
+        reactionModel.enlarge(reactEdge=True,unimolecularReact=flags,
```
Comment (alongd) [inline rmgpy/cantherm/explorer.py]: Aren't we missing the first argument, `newObject`, in the `enlarge()` call?

**Comment by alongd** [comment]
@mjohnson541, could you take a look at the Codacy report? There are many unused imports, and one instance of `i` which isn't used in a loop with `enumerate`

File: examples/cantherm/networks/methoxy/input.py [hunk 7]
```diff
@@ -221,3 +234,11 @@
 #    activeJRotor = False, #causes cantherm to crash
     rmgmode = False, 
 )
+
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply examples/cantherm/networks/methoxy/input.py]: It should run sensitivity analysis twice, once before the exploration job (when the pdep job is run) and once afterwards (when explorer reruns the pdep job at the end).  The second time should overwrite the files giving you the expected output.

File: rmgpy/rmg/model.py [hunk 8]
```diff
@@ -512,7 +512,7 @@ def makeNewReaction(self, forward, checkExisting=True):
         forward.reactants = reactants
         forward.products  = products
 
-        if checkExisting:
+        if checkExisting and hasattr(forward,'family'):
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/rmg/model.py]: Sorry, I've confirmed that was accidentally included from an earlier approach.

File: rmgpy/rmg/pdep.py [hunk 9]
```diff
@@ -318,6 +321,194 @@ def addPathReaction(self, newReaction):
             self.pathReactions.append(newReaction)
             self.invalidate()
 
+    def get_energy_filtered_reactions(self,T,tol):
+        """
+        Returns a list of products and isomers that are greater in Free Energy
+        than a*R*T + Gfsource(T)
+        """
+        R = 8.314
+        dE = tol*R*T
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/rmg/pdep.py]: They're required in the functional call so the method will throw an error if they aren't sent to the method.

File: rmgpy/cantherm/explorer.py (hunk 4 above)
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: My issue with that is that NetworkError tends to imply there's something wrong with the behavior of the network object, while the actual problem is with which networks RMG decided to generate.

File: rmgpy/cantherm/explorer.py (hunk 6 above)
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: It's optional, and RMG doesn't call it with it when it reacts

File: rmgpy/cantherm/explorer.py (hunk 5 above)
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: I dissociated that pointer more for good practice than anything else, but the speed up resulting from it is absolutely insignificant.

File: rmgpy/cantherm/explorer.py (hunk 3 above)
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: I feel like the original syntax is shorter.  I think we could conceivably have a network without any isomers.

File: documentation/source/users/cantherm/input_explorer.rst [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+pressure-dependent calculation input file you append an explorer block.  For example ::
+
+    explorer(
+        source=['methoxy'],
+        explore_tol=(1e-2,'s^-1'),
+        energy_tol=1e4,
+        flux_tol=1e-6,
+    )
+
+The ``source`` is the list of species that are part of the source channel/isomer of the network.
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply documentation/source/users/cantherm/input_explorer.rst]: added

File: documentation/source/users/cantherm/input_explorer.rst [hunk 11]
```diff
@@ -0,0 +1,55 @@
+***************************************************************
+Creating Input Files for Pressure Dependent Network Exploration
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply documentation/source/users/cantherm/input_explorer.rst]: added

File: documentation/source/users/cantherm/input_explorer.rst [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+``energy_tol`` is related to the maximum difference in E0 allowed between a reaction TS and the source channel/isomer.  
+A given reaction is deemed filterable at a given condition if ``energy_tol`` * RT < E0_TS - E0_source.  
+
+``flux_tol`` is related to the solution of a steady state problem.  In this problem a constant flux of 1.0 (choice of 
+units irrelevant) is applied to the source channel/isomer while all A => B + C reactions are assumed to be irreversible.  
+Under these conditions steady state concentrations of each isomer and the rates of every unimolecular reaction can be 
+calculated.  Since the input flux to the system is 1.0 to some degree these rates represent the fraction of the input 
+flux passing through that reaction.  When the flux through a reaction is less than ``flux_tol`` it is deemed filterable 
+at that condition.  Overall a reaction is removed from the network if it is deemed filterable by both methods at every
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply documentation/source/users/cantherm/input_explorer.rst]: done

File: rmgpy/cantherm/explorer.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        
+        if not hasattr(self.pdepjob,'outputFile'):
+            self.pdepjob.outputFile = None
+
+    def copy(self):
+        """
+        Return a copy of the pressure dependence job.
+        """
+        return ExplorerJob(
+            source=self.source,
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: I've changed this to be a deepcopy(self.source)

File: rmgpy/cantherm/explorer.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+            explore_tol=self.explore_tol,
+            energy_tol=self.energy_tol,
+            flux_tol=self.flux_tol
+        )
+        
+    def execute(self, outputFile, plot, format='pdf', print_summary=True, speciesList=None, thermoLibrary=None, kineticsLibrary=None):
+        
+        logging.info('Exploring network...')
+        
+        reactionModel = CoreEdgeReactionModel()
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: done

File: rmgpy/cantherm/data/methoxy_explore.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
+title = 'methoxy decomposition to H + CH2O'
+
+description = \
+"""
+This example illustrates how to manually set up a CanTherm input file for a small P-dep reaction system [using only the
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/data/methoxy_explore.py]: done

File: rmgpy/cantherm/data/methoxy_explore.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+    ],
+    spinMultiplicity = 2, 
+    opticalIsomers = 2,
+    collisionModel = TransportData(sigma=(3.69e-10,'m'), epsilon=(4.0,'kJ/mol')),
+    energyTransferModel = SingleExponentialDown(alpha0=(0.956,'kJ/mol'), T0=(300,'K'), n=0.95),
+)
+
+species(
+    label = 'He',
+#    freqScaleFactor = 1, # TypeError: species() got an unexpected keyword argument 'freqScaleFactor'.
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/data/methoxy_explore.py]: done

File: rmgpy/cantherm/explorer.py [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+        self.flux_tol = flux_tol
+        
+        self.pdepjob = pdepjob
+        
+        if not hasattr(self.pdepjob,'outputFile'):
+            self.pdepjob.outputFile = None
+
+    def copy(self):
+        """
+        Return a copy of the pressure dependence job.
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: done

File: rmgpy/cantherm/main.py [hunk 18]
```diff
@@ -265,11 +271,104 @@ def execute(self):
 
         # run kinetics and pdep jobs (and outputing chemkin stuff)
         for job in self.jobList:
-            if isinstance(job,KineticsJob) or isinstance(job, PressureDependenceJob):
-                job.execute(outputFile=outputFile, plot=self.plot)
-
+            if isinstance(job,KineticsJob) or isinstance(job, PressureDependenceJob) or isinstance(job, ExplorerJob):
+                if isinstance(job, ExplorerJob):
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: done

File: rmgpy/cantherm/main.py [hunk 19]
```diff
@@ -265,11 +271,104 @@ def execute(self):
...
+        name = 'kineticsjobs'
+                
+        speciesList = []
+        reactionList = []
+        for job in self.jobList:
+            if isinstance(job,ThermoJob) or isinstance(job,StatMechJob):
+                speciesList.append(job.species)
+            elif isinstance(job,KineticsJob):
+                reactionList.append(job.reaction)
+                for spc in job.reaction.reactants:
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: done

File: rmgpy/cantherm/main.py [hunk 20]
```diff
@@ -265,11 +271,104 @@ def execute(self):
...
+                    if spc.isIsomorphic(rspc):
+                        rxn.products[i] = spc
+                        break
+        del_inds = []
+        for i,spc1 in enumerate(speciesList):
+            for j,spc2 in enumerate(speciesList):
+                if j>i and spc1.isIsomorphic(spc2):
+                    del_inds.append(j)
+        
+        logging.info(del_inds)
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: yep

File: rmgpy/cantherm/main.py [hunk 21]
```diff
@@ -265,11 +271,104 @@ def execute(self):
...
+            elif isinstance(job,KineticsJob):
+                reactionList.append(job.reaction)
+                for spc in job.reaction.reactants:
+                    speciesList.append(spc)
+                for spc in job.reaction.products:
+                    speciesList.append(spc)
+        
+        speciesList = list(set(speciesList))
+        
+        logging.info('spcs:')
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: yep

File: rmgpy/cantherm/main.py [hunk 22]
```diff
@@ -265,11 +271,104 @@ def execute(self):
...
+        for i,spc1 in enumerate(speciesList):
+            for j,spc2 in enumerate(speciesList):
+                if j>i and spc1.isIsomorphic(spc2):
+                    del_inds.append(j)
+        
+        logging.info(del_inds)
+        for j in sorted(del_inds)[::-1]:
+            del speciesList[j]
+        
+        logging.info('rxns:')
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: it's debug leftovers

File: rmgpy/cantherm/main.py [hunk 23]
```diff
@@ -265,11 +271,104 @@ def execute(self):
...
+                                        shortDesc = species.thermo.comment
+               )                
+            else:
+                logging.warning('Species {0} did not contain any thermo data and was omitted from the thermo library.'.format(str(species)))
+
+        # load kinetics library entries                    
+        kineticsLibrary = KineticsLibrary(name=name,autoGenerated=True)
+        kineticsLibrary.entries = {}
+        for i in range(len(reactionList)):
+            reaction = reactionList[i]
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: done

File: rmgpy/cantherm/main.py [hunk 24]
```diff
@@ -265,11 +271,104 @@ def execute(self):
...
+            elif hasattr(reaction,'library') and reaction.library:
+                entry.longDesc = 'Originally from reaction library: ' + reaction.library + "\n" + reaction.kinetics.comment
+            else:
+                entry.longDesc = reaction.kinetics.comment
+            
+            kineticsLibrary.entries[i+1] = entry
+        
+        kineticsLibrary.label = name
+        
+        return thermoLibrary,kineticsLibrary,speciesList
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/cantherm/main.py]: done

File: rmgpy/rmg/pdep.py [hunk 25]
```diff
@@ -318,6 +321,194 @@ def addPathReaction(self, newReaction):
             self.pathReactions.append(newReaction)
             self.invalidate()
 
+    def get_energy_filtered_reactions(self,T,tol):
+        """
+        Returns a list of products and isomers that are greater in Free Energy
+        than a*R*T + Gfsource(T)
+        """
+        R = 8.314
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/rmg/pdep.py]: done

File: rmgpy/rmg/pdep.py [hunk 26]
```diff
@@ -318,6 +321,194 @@ def addPathReaction(self, newReaction):
...
+        for conf in self.isomers+self.products+self.reactants:
+            if len(conf.species) == len(self.source):
+                if len(self.source) == 1:
+                    if self.source[0].isIsomorphic(conf.species[0]):
+                        E0source = conf.E0
+                        break
+                elif len(self.source) == 2:
+                    boo00 = self.source[0].isIsomorphic(conf.species[0])
+                    boo01 = self.source[0].isIsomorphic(conf.species[1])
+                    boo10 = self.source[1].isIsomorphic(conf.species[0])
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/rmg/pdep.py]: done

File: environment_mac.yml [hunk 27]
```diff
@@ -40,3 +40,4 @@ dependencies:
   - mock
   # - muq
   - pymongo
+  - mpmath
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply environment_mac.yml]: done

File: environment_windows.yml [hunk 28]
```diff
@@ -39,3 +39,4 @@ dependencies:
   - mock
   # - muq
   - pymongo
+  - mpmath
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply environment_windows.yml]: done

File: documentation/source/users/cantherm/input_explorer.rst [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+added to the network and reacted and the loop is flagged to cycle through all of the temperatures and pressures
+again.  Once this loop is finished a network_full.py file is generated in the pdep directory that has the full
+explored network.  
+
+Network Reduction
+=================
+
+``energy_tol`` and ``flux_tol`` control reduction of the full network.  This is highly recommended as the full model
+will often contain many unimportant reactions.  How strict you decide to set these parameters should be related
+to how much you trust the thermo and rate information being used in the job.
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply documentation/source/users/cantherm/input_explorer.rst]: done

File: rmgpy/rmg/pdep.py [hunk 30]
```diff
@@ -318,6 +321,194 @@ def addPathReaction(self, newReaction):
...
+            else:
+                ind = None
+            if rxn.products[0] in isomerSpcs:
+                ind2 = isomerSpcs.index(rxn.products[0])
+                kr = rxn.getRateCoefficient(T,P)/rxn.getEquilibriumConstant(T)
+                A[ind2,ind2] -= kr
+            else:
+                ind2 = None
+            
+            if ind is not None and ind2 is not None:
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/rmg/pdep.py]: done

File: rmgpy/rmg/pdep.py [hunk 31]
```diff
@@ -318,6 +321,194 @@ def addPathReaction(self, newReaction):
...
+                A[ind,ind2] += kr
+                A[ind2,ind] += kf
+            
+            if bimolecular:
+                if rxn.reactants[0].species == self.source:
+                    if not ind:
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/rmg/pdep.py]: done

File: rmgpy/rmg/pdep.py (hunk 1 above)
Comment (mjohnson541) [reply rmgpy/rmg/pdep.py]: fixed

File: rmgpy/cantherm/explorer.py (hunk 2 above)
Comment (mjohnson541) [reply rmgpy/cantherm/explorer.py]: fixed

**Comment by mjohnson541** [comment]
Ok, I think I've made all of those changes and some of the codacy ones.

File: rmgpy/rmg/model.py [hunk 32]
```diff
@@ -512,7 +512,7 @@ def makeNewReaction(self, forward, checkExisting=True):
         forward.reactants = reactants
         forward.products  = products
 
-        if checkExisting:
+        if checkExisting: #and hasattr(forward,'family'):
```
Comment (alongd) [inline rmgpy/rmg/model.py]: Perhaps remove the comment?

File: rmgpy/cantherm/explorer.py [hunk 33]
```diff
@@ -0,0 +1,232 @@
...
+        self.pdepjob.execute(outputFile, plot, format='pdf', print_summary=True)
+        
+        if warns != []:
+            logging.info('\nOUTPUT WARNINGS:\n')
+            for w in warns:
+                logging.warning(w)
+
+
+        
+
```
Comment (alongd) [inline rmgpy/cantherm/explorer.py]: The EOF still looks strange. Can you remove the tabs/spaces and leave just one empty line at the end?

File: examples/cantherm/networks/methoxy/input.py [hunk 34]
```diff
@@ -221,3 +234,11 @@
...
+explorer(
+	source=['methoxy'],
+	explore_tol=(1e-2,'s'),
```
Comment (alongd) [inline examples/cantherm/networks/methoxy/input.py]: Should the unit be `s^-1`?

**Comment by alongd** [comment]
Testing the explorer with some arbitrary `reactionLibraries` gives:
```
Traceback (most recent call last):
  File "/home/alongd/Code/RMG-Py//cantherm.py", line 60, in <module>
    cantherm.execute()
  File "/home/alongd/Code/RMG-Py/rmgpy/cantherm/main.py", line 233, in execute
    self.jobList = self.loadInputFile(self.inputFile)
  File "/home/alongd/Code/RMG-Py/rmgpy/cantherm/main.py", line 215, in loadInputFile
    self.jobList = loadInputFile(self.inputFile)
  File "/home/alongd/Code/RMG-Py/rmgpy/cantherm/input.py", line 482, in loadInputFile
    exec f in global_context, local_context
  File "input.py", line 8, in <module>
    kineticsEstimator = 'rate rules',
  File "/home/alongd/Code/RMG-Py/rmgpy/cantherm/input.py", line 143, in database
    depository = False, # Don't bother loading the depository information, as we don't use it
  File "/home/alongd/Code/RMG-Py/rmgpy/data/rmg.py", line 104, in load
    kineticsDepositories
  File "/home/alongd/Code/RMG-Py/rmgpy/data/rmg.py", line 171, in loadKinetics
    depositories=kineticsDepositories
  File "/home/alongd/Code/RMG-Py/rmgpy/data/kinetics/database.py", line 106, in load
    self.loadLibraries(os.path.join(path, 'libraries'), libraries)
  File "/home/alongd/Code/RMG-Py/rmgpy/data/kinetics/database.py", line 225, in loadLibraries
    library_file = os.path.join(path, library_name,'reactions.py')
  File "/home/alongd/anaconda2/envs/rmg_env/lib/python2.7/posixpath.py", line 68, in join
    if b.startswith('/'):
AttributeError: 'tuple' object has no attribute 'startswith'

```

**Comment by mjohnson541** [comment]
Ok, that issue should be fixed.  Furthermore, now the pdep job isn't run until after the network is generated and if the bathGas is specified in the explorer block a network block doesn't need to be specified.

**Comment by mjohnson541** [comment]
I've added the ability to automatically estimate kinetics and thermo using RMG's database

**Comment by mjohnson541** [comment]
Ok, I've fixed the testing problems with this branch.  Is this ready to merge now?

File: environment_windows.yml (hunk 28 above)
[the remark this answers is not in the document]
Comment (mliu49) [reply environment_windows.yml]: FYI, the last line is not empty (contains spaces).

**Comment by alongd** [comment]
Where is `maximumRadicalElectrons` being set? Can you add it to the example?

**Comment by mjohnson541** [comment]
It's set in the explorer block, I'll add it to the example.

**Comment by mjohnson541** [comment]
The species constraints bug is fixed now and the maximumRadicalElectrons option has been added to the methoxy example.

**Comment by mjohnson541** [comment]
The issues with your N2H4 test case (with the fixBarrierHeight issue) are now fixed.

**Comment by mjohnson541** [comment]
Ok, I've improved the "can't find source" error message to suggest that users check their libraries don't have reactions that are not flagged elementary_high_p blocking the RMG generated reactions.  Is this ready to merge or did you want to test it some more?

**Pull request merged**